core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,328 @@
+      if (tts!!.stop() == TextToSpeech.SUCCESS) {
+        tts!!.setOnUtteranceProgressListener(null)
+        onSpeakingListener.onSpeakingEnded()
+      }
+    } else {
+      val locale = iSO3ToLocale(zimReaderContainer.language)
+      var result = 0
+      if ("mul" == zimReaderContainer.language) {
+        Log.d(
+          TAG_KIWIX, "TextToSpeech: disabled " +
+            zimReaderContainer.language
+        )
+        Toast.makeText(context, R.string.tts_not_enabled, Toast.LENGTH_LONG).show()
+        return
+      }
+      if (locale == null || tts?.isLanguageAvailable(locale)
+          .also {
+            it?.let {
+              result = it
+            }
+          } == TextToSpeech.LANG_MISSING_DATA || result == TextToSpeech.LANG_NOT_SUPPORTED
+      ) {
+        Log.d(
+          TAG_KIWIX, "TextToSpeech: language not supported: " +
+            zimReaderContainer.language
+        )
+        Toast.makeText(context, R.string.tts_lang_not_supported, Toast.LENGTH_LONG).show()
+      } else {
+        tts?.language = locale
+        if (getFeatures(tts, locale).contains(TextToSpeech.Engine.KEY_FEATURE_NOT_INSTALLED)) {
+          context.toast(R.string.tts_lang_not_supported, Toast.LENGTH_LONG)
+          return
+        }
+        if (requestAudioFocus()) {
+          loadURL(webView)
+        }
+      }
+    }
+  }
+
+  @SuppressLint("NewApi") private fun getFeatures(
+    tts: TextToSpeech?,
+    locale: Locale
+  ): Set<String> {
+    return tts!!.voice
+      .features
+  }
+
+  private fun loadURL(webView: WebView) {
+    // We use JavaScript to get the content of the page conveniently, earlier making some
+    // changes in the page
+    webView.loadUrl(
+      """
+        javascript:body = document.getElementsByTagName('body')[0].cloneNode(true);
+        // Remove some elements that are shouldn't be read (table of contents,
+        // references numbers, thumbnail captions, duplicated title, etc.)
+        "toRemove = body.querySelectorAll('sup.reference, #toc, .thumbcaption, 
+        "    title, .navbox');
+        "Array.prototype.forEach.call(toRemove, function(elem) {
+        "    elem.parentElement.removeChild(elem);
+        "});
+        "tts.speakAloud(body.innerText);
+        """
+    )
+  }
+
+  fun stop() {
+    tts?.let { tts ->
+      {
+        if (tts.stop() == TextToSpeech.SUCCESS) {
+          currentTTSTask = null
+          tts.setOnUtteranceProgressListener(null)
+          onSpeakingListener.onSpeakingEnded()
+        }
+      }
+    }
+  }
+
+  private fun requestAudioFocus(): Boolean {
+    val audioFocusRequest = am.requestAudioFocus(
+      onAudioFocusChangeListener, AudioManager.STREAM_MUSIC,
+      AudioManager.AUDIOFOCUS_GAIN
+    )
+    Log.d(TAG_KIWIX, "Audio Focus Requested")
+    synchronized(focusLock) {
+      return@requestAudioFocus audioFocusRequest ==
+        AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    }
+  }
+
+  fun pauseOrResume() {
+    if (currentTTSTask == null) {
```

Comment:
this is a good place for `let` because you only want to do stuff when the task is not null

```
currentTask?.let{ task ->
if(task.isPaused)
...etc
}
```

Reply:
@macgills I changed that.